### PR TITLE
cmd: Add Rust-Postgres to Nightlies:

### LINKED ISSUE #48627: roachtest: add rust-postgres driver tests to nightlies
This driver seems to have a robust test suite that verifies lots of pgwire compatibility. See https://github.com/cockroachdb/cockroach/issues/48448

Jira issue: CRDB-4307

### PR DESCRIPTION
**Pull request by e-mbrown**
Resolves #48627
Add new roachtest for Rust Postgres

Release note: None

Jira issue: CRDB-14717

### CODE REVIEW

File: pkg/cmd/roachtest/tests/rust_helpers.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+	for scanner.Scan() {
+		x := scanner.Text()
+		match := rustUnitTestOutputRegex.FindStringSubmatch(x)
+		if match != nil {
+			groups := map[string]string{}
+			for i, name := range match {
+				groups[rustUnitTestOutputRegex.SubexpNames()[i]] = name
+			}
+			test := fmt.Sprintf("%s %s::%s", groups["type"], groups["class"], groups["name"])
+			skipped := groups["result"] == "skipped" || groups["result"] == "expected failure" || len(groups["result"]) > 4
```
Comment (otan) [inline pkg/cmd/roachtest/tests/rust_helpers.go:25]: can you explain the `len(groups["result"]) > 4` thing?

File: pkg/cmd/roachtest/tests/rust_helpers.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"regexp"
+)
+
+var rustUnitTestOutputRegex = regexp.MustCompile(`(?P<type>.*) (?P<class>.*)::(?P<name>.*) .\.\.\ (?P<result>.*)`)
+
+func (r *ormTestsResults) parseRustUnitTestOutput(
```
Comment (otan) [inline pkg/cmd/roachtest/tests/rust_helpers.go:12]: this looks very similar to `parsePythonUnitTestOutput` barring a few words. is it possible to add parameters and make them both reference the same impl,

e.g.

```

func (r *ormTestsResults) parseUnitTestOutput(input []byte, expectedFailures blocklist, ignoreList blocklist, outputRe *regexp.Regexp, resultName string) {
   // replace ["reason"] and rustUnitTestOutputRegex here
}

func (r *ormTestsResults) parseRustUnitTestOutput(input []byte, expectedFailures blocklist, ignoreList blocklist) {
   return r.parseUnitTestOutput(input, expectedFailure, ignoreList, re, resultName)
}

func (r *ormTestsResults) parsePythonUnitTestOutput(input []byte, expectedFailures blocklist, ignoreList blocklist) {
   return r.parseUnitTestOutput(input, expectedFailure, ignoreList, re, resultName)
}
```

if you feel there's more differences than that, please feel free to skip this suggestion.

File: pkg/cmd/roachtest/tests/rust_helpers.go (hunk 1 above)
Comment (e-mbrown) [reply pkg/cmd/roachtest/tests/rust_helpers.go]: That was left over from testing. Removed it

File: pkg/cmd/roachtest/tests/rust_postgres.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/option"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/registry"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/spec"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test"
+	"github.com/cockroachdb/cockroach/pkg/roachprod/install"
+)
+
+func registerRustPostgres(r registry.Registry) {
+	runRustPostgres := func(ctx context.Context, t test.Test, c cluster.Cluster) {
+		// Ask about this line
```
Comment (otan) [inline pkg/cmd/roachtest/tests/rust_postgres.go:27]: what did you want to ask

File: pkg/cmd/roachtest/tests/rust_postgres.go [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+		}
+
+		if err := c.RunE(
+			ctx,
+			node,
+			"cd /mnt/data1 && git clone https://github.com/sfackler/rust-postgres.git",
+		); err != nil {
+			t.Fatal(err)
+		}
+		// TODO: Check if one is actually the option we want, may be 2
```
Comment (otan) [inline pkg/cmd/roachtest/tests/rust_postgres.go:64]: what may be 2?

**Review by otan — COMMENTED** [review]
nit: can you remove the trialing `:` from the commit title and the PR description?

File: pkg/cmd/roachtest/tests/rust_postgres.go (hunk 4 above)
Comment (e-mbrown) [reply pkg/cmd/roachtest/tests/rust_postgres.go:64]: Oop, this is another leftover comment from testing. Removing

File: pkg/cmd/roachtest/tests/rust_postgres.go (hunk 3 above)
Comment (e-mbrown) [reply pkg/cmd/roachtest/tests/rust_postgres.go:27]: Why cant roachtest be run in local mode?

File: pkg/cmd/roachtest/tests/parsing_helpers.go [hunk 5]
```diff
@@ -17,18 +17,31 @@ import (
...
+}
+
+func (r *ormTestsResults) parseRustUnitTestOutput(
+	input []byte, expectedFailures blocklist, ignoredList blocklist,
+) {
+	r.parseUnitTestOutput(rustUnitTestOutputRegex, input, expectedFailures, ignoredList)
+}
+
+func (r *ormTestsResults) parseUnitTestOutput(
+	regex *regexp.Regexp, input []byte, expectedFailures blocklist, ignoredList blocklist,
```
Comment (otan) [inline pkg/cmd/roachtest/tests/parsing_helpers.go:36]: nit: rename `regex` to  `testOutputRegex`

File: pkg/cmd/roachtest/tests/rust_postgres.go [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+		); err != nil {
+			t.Fatal(err)
+		}
+
+		t.Status("installing Docker")
+		if err := c.Install(ctx, t.L(), c.All(), "docker"); err != nil {
+			t.Fatal(err)
+		}
+
+		const dockerRun = `sudo docker run -p 5433:26257 --name rust -d cockroachdb/cockroach:latest start-single-node --insecure`
```
Comment (otan) [inline pkg/cmd/roachtest/tests/rust_postgres.go:103]: so this command doesn't run the cockroachdb binary that comes along with `roachtest`, e.g. `roachtest --binary`. we should use the raw SQL command instead. see above.

File: pkg/cmd/roachtest/tests/rust_postgres.go [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+
+func registerRustPostgres(r registry.Registry) {
+	runRustPostgres := func(ctx context.Context, t test.Test, c cluster.Cluster) {
+		if c.IsLocal() {
+			t.Fatal("can not be run in local mode")
+		}
+		node := c.Node(1)
+		t.Status("setting up cockroach")
+		c.Put(ctx, t.Cockroach(), "./cockroach", c.All())
+		c.Start(ctx, t.L(), option.DefaultStartOpts(), install.MakeClusterSettings(), c.All())
```
Comment (otan) [inline pkg/cmd/roachtest/tests/rust_postgres.go:33]: if we need this to be port 5433, we should add an option to https://github.com/cockroachdb/cockroach/blob/5af48f0e85f10124161679408024b29623d78f73/pkg/roachprod/install/cockroach.go#L105 and make it take in 5433 as an arg, and augment `--port` to this port number. alternatively, can we change the PGURL to use 26257 instead?

File: pkg/cmd/roachtest/tests/rust_postgres.go [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+		}
+		t.L().Printf("%s", status)
+
+		t.Status("Running rust-postgres test suite")
+
+		result, err := c.RunWithDetailsSingleNode(
+			ctx,
+			t.L(),
+			node,
+			`cd /mnt/data1/rust-postgres && PGPORT=5433 /home/ubuntu/.cargo/bin/cargo test 2>&1 > rustpostgres.stdout --no-fail-fast`)
```
Comment (otan) [inline pkg/cmd/roachtest/tests/rust_postgres.go:137]: does 5433 work? should this be 26257?

File: pkg/cmd/roachtest/tests/rust_postgres.go [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+			t,
+			c,
+			node,
+			" Installing more build essentials",
+			"sudo apt-get install -y pkg-config libssl-dev",
+		); err != nil {
+			t.Fatal(err)
+		}
+
+		//t.Status("installing Docker")
```
Comment (otan) [inline pkg/cmd/roachtest/tests/rust_postgres.go:102]: let's delete these?

File: pkg/cmd/roachtest/tests/rust_postgres.go (hunk 8 above)
Comment (otan) [reply pkg/cmd/roachtest/tests/rust_postgres.go:124]: does that map to 26257 somehow? how does it connect to cockroach?

File: pkg/cmd/roachtest/tests/parsing_helpers.go (hunk 5 above)
Comment (otan) [reply pkg/cmd/roachtest/tests/parsing_helpers.go:36]: ping on this comment

**Review by otan — APPROVED** [review]

File: pkg/cmd/roachtest/tests/rust_postgres.go [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+		if ignorelist != nil {
+			status = fmt.Sprintf(
+				"Running cockroach %s, using blocklist %s, using ignorelist %s",
+				version,
+				blocklistName,
+				ignorelistName)
+		}
+		t.L().Printf("%s", status)
+
+		c.Stop(ctx, t.L(), option.DefaultStopOpts(), c.All())
```
Comment (otan) [inline pkg/cmd/roachtest/tests/rust_postgres.go:119]: can you add a comment here before merging about why we do this?

**Comment by e-mbrown** [comment]
bors r=otan

**Comment by e-mbrown** [comment]
bors r=otan

**Pull request merged**